html/semantics/embedded-content/the-iframe-element/iframe-loading-lazy-history-pushState.html

```diff
@@ -0,0 +1,33 @@
+<!DOCTYPE html>
+<title>History state change for iframe loading='lazy' before it is loaded: history.pushState</title>
+<iframe data-src="about:blank#foo" src="support/blank.htm?src" loading="lazy"></iframe>
+<script>
+const iframe = document.querySelector('iframe');
+let pushStateSuccess = true;
+try {
+  // Should have no effect on lazy-loading.
+  // Per https://html.spec.whatwg.org/#can-have-its-url-rewritten
+  // only the fragment can be changed for about: URLs.
+  iframe.contentWindow.history.pushState(null, "", iframe.dataset.src);
+} catch(ex) {
+  replaceStateSuccess = false;
+}
+const locationAfterPushState = iframe.contentWindow.location.href;
+</script>
+<!-- Loading testharness.js here is intentional to reproduce a bug in WebKit. -->
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script>
+setup({single_test: true});
+assert_true(pushStateSuccess);
+assert_equals(locationAfterPushState, new URL("about:blank#foo", location.href).href);
+onload = () => {
+  step_timeout(() => {
+    assert_equals(iframe.contentWindow.location.href, new URL("support/blank.htm?src", location.href).href);
+    done();
+  }, 1000);
```

Comment:
Does this have to be done on a timeout? Or can it be done on the lazyloaded iframe's load event, since it will be fired asynchronously (with respect to this code currently running) once `support/blank.html` loads?

Reply:
Since some tests do potentially two navigations, a timeout is still needed. But maybe I can use the iframe load event instead of the window load event...

Reply:
>> Or can it be done on the lazyloaded iframe's load event, since it will be fired asynchronously
>
> But maybe I can use the iframe load event instead of the window load event...

So the latest commits use the iframe's load event instead of the window load event. Does that mean we can remove the timeout now?

Reply:
No, since the bug is about the lazy navigation happens after the other navigation, removing the timeout could mean that we check the state of the world between those two navigations. With some more complexity, it's possible to detect the second `load` event to fail the test earlier, but it doesn't help the conforming case.

Reply:
OK makes sense, maybe if you could just document the intent of the timeout that would be great! Thank you.

Reply:
Done, thanks

Reply:
I'm sorry I want to revisit this. Maybe I've just paged all of the useful context out of my brain, but remind me what are the "two navigations"? There are two interesting things that happen:
 1. Creation of the initial about:blank Document when the iframe is inserted
 2. Process the iframe attributes: set the lazyload resumption steps to eventually resume the `src`-attribute navigation
 3. Push state on the initial about:blank Document
 4. Unhide the iframe, thus resuming the lazyloaded `src`-attribute navigation

The only real _navigation_ that happens here is (4), so I'm still a tad confused where step_timeout comes into play here. Sorry.

Reply:
Fixed in 89e288113394a018426c6c3badc4ebf4c37476fb

Reply:
My lingering thought on this is I wonder if in the test cases where there are two possible navigations, if a single zero-ms timeout is enough? That is, maybe it is still possible that the timeout is short enough such that even after it, you could still observe the state in between the two navigations (i.e., the timeout fires before the second navigation is actually finished, even though there shouldn't be a second navigation at all).

If you don't think that's possible I'll leave this alone though. The tests generally look good (still reviewing just figured I'd leave this comment here)

Reply:
The navigation depends on a fetch completing which could take some time (but usually it's fast). From prior experience, 2s is generally needed to avoid flaky results.

Reply:
Yep, makes sense. Please see my edit above where I was mistaken about the timeout length.